src/ucs/sys/ucs_rtnetlink.c

```diff
@@ -0,0 +1,135 @@
+/**
+* Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2024. ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <ucs/debug/log.h>
+#include <ucs/debug/memtrack_int.h>
+#include <ucs/sys/ucs_rtnetlink.h>
+#include <ucs/sys/ucs_netlink.h>
+#include <ucs/sys/sock.h>
+#include <ucs/type/status.h>
+
+#include <errno.h>
+#include <linux/rtnetlink.h>
+
+#define NETLINK_MESSAGE_MAX_SIZE 8195
+
+
+struct ucs_netlink_route_info_t {
+    const struct sockaddr *sa_remote;
+    int                    if_index;
+    int                    matching;
+};
+
+static ucs_status_t ucs_rtnetlink_get_route_info(struct rtattr *rta, int len,
+                                                 int *if_idx, void **dst_in_addr)
+{
+    *if_idx      = -1;
+    *dst_in_addr = NULL;
+
+    for (; RTA_OK(rta, len); rta = RTA_NEXT(rta, len)) {
+        if (rta->rta_type == RTA_OIF) {
+            *if_idx = *((int *)RTA_DATA(rta));
+        } else if (rta->rta_type == RTA_DST) {
+            *dst_in_addr = RTA_DATA(rta);
+        }
+    }
+
+    if ((*if_idx == -1) || (*dst_in_addr == NULL)) {
+        ucs_diag("either iface index or dest address are missing "
+                  "in the routing table entry");
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    return UCS_OK;
+}
+
+static int ucs_rtnetlink_is_rule_matching(struct rtmsg *rtm, size_t rtm_len,
+                                          const struct sockaddr *sa_remote,
+                                          int iface_index)
+{
+    int   rule_iface;
+    void *dst_in_addr;
+
+    if (ucs_rtnetlink_get_route_info(RTM_RTA(rtm), rtm_len, &rule_iface,
+                                     &dst_in_addr) != UCS_OK) {
+        return 0;
+    }
+
+    if (rule_iface != iface_index) {
+        return 0;
+    }
+
+    return ucs_bitwise_is_equal(ucs_sockaddr_get_inet_addr(sa_remote),
+                                dst_in_addr, rtm->rtm_dst_len);
+}
+
+static ucs_status_t
+ucs_netlink_parse_rt_entry_cb(struct nlmsghdr *nlh, void *nl_msg, void *arg)
+{
+    struct ucs_netlink_route_info_t *info = (struct ucs_netlink_route_info_t *)arg;
+
+    if (ucs_rtnetlink_is_rule_matching((struct rtmsg *)nl_msg, RTM_PAYLOAD(nlh),
+                                       info->sa_remote, info->if_index)) {
+        info->matching = 1;
+        return UCS_OK;
+    }
+
+    return UCS_INPROGRESS;
+}
+
+int ucs_netlink_rule_exists(const char *iface_name,
+                            const struct sockaddr *sa_remote)
+{
+    char *recv_msg                       = NULL;
+    struct ucs_netlink_route_info_t info = {0};
+    struct rtmsg rtm                     = {0};
+    ucs_status_t status;
+    size_t recv_msg_len;
+    int iface_index;
+
+    iface_index = if_nametoindex(iface_name);
```

Comment:
can we have some sort of caching so that if is_reachable is called multiple times, even to different destinations, we will not send netlink command every time?
seems we could read the routing table just once?